# stMMR: Multi-Modal Feature Representation in Spatial Transcriptomics with Similarity Contrastive Learning

## Overview
Understanding the spatial distribution and function of cells from spatial transcriptomics (ST) is of great value for deciphering the development and differentiation of tissues. However, the inherent significant heterogeneity and varying spatial resolutions of spatial transcriptomics multimodal data present challenges in the joint analysis of these modalities. In this study, we propose a novel method, named stMMR, to achieve multi-modal feature representation based on similarity contrastive learning. stMMR integrates histological imaging information with gene expression data through adjacency relationships. It uses self-attention module for deep embedding of features within a modality and incorporates similarity contrastive learning for integrating features across modalities. stMMR demonstrates superior performances in multiple analyses using datasets generated by different platforms, including domain identification, developmental trajectory inference as well as enhancement of gene expression. Using stMMR, we systematically analyzed the developmental process of the chicken heart and conducted an in-depth examination of key genes in breast cancer and lung cancer. In conclusion, stMMR is capable of effectively integrating the multimodal information and exhibits superior adaptability as well as stability in handling different types of ST data. 


![](https://github.com/nayu0419/stMMR/blob/main/overview.png)

## Installation

Clone this repository. The stMMR has been implemented in Python3.9.1 and Pytorch 1.11.1.
### Github
```
git clone https://github.com/nayu0419/stMMR.git
cd stMMR
python setup.py build
python setup.py install --user
```
### PyPI: Directly install the package from PyPI
```
pip3 install stMMR
```


## Requirements
* numpy~=1.21.5
* numba~=0.55.1
* scanpy~=1.9.3
* torch~=1.11.0
* munkres~=1.1.4
* pandas~=1.3.5
* scikit-learn~=1.0.2
* anndata~=0.8.0
* scipy~=1.7.3
* matplotlib~=3.5.2

## Datasets
All datasets used in our paper can be found in:

* DPLFC: 
The primary source: https://github.com/LieberInstitute/spatialLIBD; 
The processed version: https://www.nature.com/articles/s41593-020-00787-0.

* Human breast cancer: 
The primary source: https://www.10xgenomics.com/resources/datasets/human-breast-cancer-block-a-section-1-1-standard-1-1-0; 
The processed version: https://github.com/JinmiaoChenLab/SEDR_analyses/.

* Chicken heart
The primary source: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE149457;

* Lung cancer (9-1) nanostring
The primary source: https://nanostring.com/products/cosmx-spatial-molecular-imager/nsclc-ffpe-dataset;

* Human pancreatic ductal adenocarcinoma
The primary source: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM3036911;

* Mouse anterior brain
The primary source: https://www.10xgenomics.com/resources/datasets/mouse-brain-serial-section-1-sagittal-anterior-1-standard-1-1-0;




## Example usage

```
'python run_Breast_Cancer.py'
```

## Contact
Feel free to submit an issue or contact us at gaorui@sdu.edu.cn for problems about the packages.


